nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py

```diff
@@ -165,6 +170,13 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
             if not parallel_state.is_pipeline_last_stage():
                 output_tensor = output_tensor.to(dtype=self.autocast_dtype)
 
+            def logprobs_func(output_tensor, non_loss_data=True):
```

Comment:
Why is there a `non_loss_data=True` here which is not used, but the standard `loss_func` doesn't have this param?

Reply:
Shouldn't we then add this to `loss_func` as well?

Reply:
If `collect_non_loss_data = False`, which is true in `loss_func`, this argument `non_loss_data ` is not called.

https://github.com/NVIDIA/Megatron-LM/blob/0bc3547702464501feefeb5523b7a17e591b21fa/megatron/core/pipeline_parallel/schedules.py#L214

Reply:
I'm also a bit confused (like @trias702): @gshennvm is saying that we use `collect_non_loss_data=True` but @shengyangs is saying that `collect_non_loss_data = False`...

Reply:
Also, `logprobs_func` doesn't seem to actually use `collect_non_loss_data` at all inside its function, despite having it in the signature?

Reply:
So I had a closer look to better understand what's going on:
* We use `collect_non_loss_data=True` to [get the ref policy log probs](https://github.com/NVIDIA/NeMo-Aligner/blob/dd6c271dfe7eb9831d3075417a4bd3f163c23a3c/nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py#L419), using `logprobs_only=True` which means we return the `logprobs_func()` function, which Megatron-LM [will call](https://github.com/NVIDIA/Megatron-LM/blob/0bc3547702464501feefeb5523b7a17e591b21fa/megatron/core/pipeline_parallel/schedules.py#L228) with `non_loss_data=True`
* We use `collect_non_loss_data=False` (default) to [compute training loss](https://github.com/NVIDIA/NeMo-Aligner/blob/dd6c271dfe7eb9831d3075417a4bd3f163c23a3c/nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py#L312), using `logprobs_only=False` which means we return the `loss_func()` function, which Megatron-LM [will call](https://github.com/NVIDIA/Megatron-LM/blob/0bc3547702464501feefeb5523b7a17e591b21fa/megatron/core/pipeline_parallel/schedules.py#L215) without providing the `non_loss_data` argument.

I think what's missing is something like this in `logprobs_func()`:
```python
def logprobs_func(output_tensor, non_loss_data=True):
    # This function is expected to be used only when `collect_non_loss_data=True` in Megatron-LM.
    # See https://github.com/NVIDIA/Megatron-LM/blob/0bc3547702464501feefeb5523b7a17e591b21fa/megatron/core/pipeline_parallel/schedules.py#L228
    assert non_loss_data
    logprobs = ...
```